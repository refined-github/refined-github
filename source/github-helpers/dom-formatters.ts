import zipTextNodes from 'zip-text-nodes';
import linkifyURLsCore from 'linkify-urls';
import linkifyIssuesCore from 'linkify-issues';

import {getRepo} from '.';
import getTextNodes from '../helpers/get-text-nodes';
import parseBackticksCore from './parse-backticks';

// Shared class necessary to avoid also shortening the links
export const linkifiedURLClass = 'rgh-linkified-code';

// If we are not in a repo, relative issue references won't make sense
// but `user`/`repo` need to be set to avoid breaking errors in `linkify-issues`
// https://github.com/sindresorhus/refined-github/issues/1305
const currentRepo = getRepo()!;

export function linkifyIssues(
	element: Element,
	options: Partial<linkifyIssuesCore.TypeDomOptions> = {}
): void {
	const linkified = linkifyIssuesCore(element.textContent!, {
		user: currentRepo.owner ?? '/',
		repository: currentRepo.name ?? '/',
		type: 'dom',
		baseUrl: '',
		...options,
		attributes: {
			rel: 'noreferrer noopener',
			class: linkifiedURLClass, // Necessary to avoid also shortening the links
			...options.attributes
		}
	});
	if (linkified.children.length === 0) { // Children are <a>
		return;
	}

	// Enable native issue title fetch
	for (const link of linkified.children as HTMLCollectionOf<HTMLAnchorElement>) {
		const issue = link.href.split('/').pop();
		link.setAttribute('class', 'issue-link js-issue-link tooltipped tooltipped-ne');
		link.dataset.errorText = 'Failed to load issue title';
		link.dataset.permissionText = 'Issue title is private';
		link.dataset.url = link.href;
		link.dataset.id = `rgh-issue-${issue!}`;
	}

	zipTextNodes(element, linkified);
}

export function linkifyURLs(element: Element): void {
	if (element.textContent!.length < 15) { // Must be long enough for a URL
		return;
	}

	const linkified = linkifyURLsCore(element.textContent!, {
		type: 'dom' as const,
		attributes: {
			rel: 'noreferrer noopener',
			class: linkifiedURLClass // Necessary to avoid also shortening the links
		}
	});

	if (linkified.children.length === 0) { // Children are <a>
		return;
	}

	zipTextNodes(element, linkified);
}

export function parseBackticks(element: Element): void {
	for (const node of getTextNodes(element)) {
		const fragment = parseBackticksCore(node.textContent!);

		if (fragment.children.length > 0) {
			node.replaceWith(fragment);
		}
	}
}
