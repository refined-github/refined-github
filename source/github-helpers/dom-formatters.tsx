import React from 'dom-chef';
import select from 'select-dom';
import zipTextNodes from 'zip-text-nodes';
import {applyToLink} from 'shorten-repo-url';
import linkifyURLsCore from 'linkify-urls';
import linkifyIssuesCore, {type TypeDomOptions} from 'linkify-issues';

import getTextNodes from '../helpers/get-text-nodes.js';
import parseBackticksCore from './parse-backticks.js';
import {buildRepoURL} from './index.js';

// Shared class necessary to avoid also shortening the links
export const linkifiedURLClass = 'rgh-linkified-code';
export const linkifiedURLSelector = '.rgh-linkified-code';

export const codeElementsSelector = [
	// Sometimes formatted diffs are loaded later and discard our formatting #5870
	'.blob-code-inner:not(deferred-diff-lines.awaiting-highlight *)', // Code lines
	':not(.notranslate) > .notranslate', // Code blocks in comments. May be wrapped twice
];

export function shortenLink(link: HTMLAnchorElement): void {
	// Exclude the link if the closest element found is not `.comment-body`
	// This avoids shortening links in code and code suggestions, but still shortens them in review comments
	// https://github.com/refined-github/refined-github/pull/4759#discussion_r702460890
	if (link.closest(String([...codeElementsSelector, '.comment-body']))?.classList.contains('comment-body')) {
		applyToLink(link, location.href);
	}
}

export function linkifyIssues(
	currentRepo: {owner?: string; name?: string},
	element: Element,
	options: Partial<TypeDomOptions> = {},
): void {
	const linkified = linkifyIssuesCore(element.textContent, {
		user: currentRepo.owner ?? '/',
		repository: currentRepo.name ?? '/',
		type: 'dom',
		baseUrl: '',
		...options,
		attributes: {
			class: linkifiedURLClass, // Necessary to avoid also shortening the links
			...options.attributes,
		},
	});
	if (linkified.children.length === 0) { // Children are <a>
		return;
	}

	// Enable native issue title fetch
	for (const link of linkified.children as HTMLCollectionOf<HTMLAnchorElement>) {
		const issue = link.href.split('/').pop();
		link.setAttribute('class', 'issue-link js-issue-link');
		link.dataset.errorText = 'Failed to load title';
		link.dataset.permissionText = 'Title is private';
		link.dataset.url = link.href;
		link.dataset.id = `rgh-issue-${issue!}`;
		link.dataset.hovercardType = 'issue';
		link.dataset.hovercardUrl = `${link.pathname}/hovercard`;
	}

	zipTextNodes(element, linkified);
}

export function linkifyURLs(element: Element): Element[] | void {
	if (element.textContent.length < 15) { // Must be long enough for a URL
		return;
	}

	if (select.exists(linkifiedURLSelector, element)) {
		return select.all(linkifiedURLSelector, element);
	}

	const linkified = linkifyURLsCore(element.textContent, {
		type: 'dom' as const,
		attributes: {
			rel: 'noreferrer noopener',
			class: linkifiedURLClass, // Necessary to avoid also shortening the links
		},
	});

	if (linkified.children.length === 0) { // Children are <a>
		return;
	}

	zipTextNodes(element, linkified);
}

export function parseBackticks(element: Element): void {
	for (const node of getTextNodes(element)) {
		const fragment = parseBackticksCore(node.textContent);

		if (fragment.children.length > 0) {
			node.replaceWith(fragment);
		}
	}
}

export function linkifyCommit(sha: string): JSX.Element {
	// Data attributes copied from the commit in https://github.com/refined-github/github-url-detection/releases/tag/v7.1.2
	return (
		<code>
			<a
				className="Link--secondary"
				href={buildRepoURL('commit', sha)}
				data-hovercard-type="commit"
				data-hovercard-url={buildRepoURL('commit', sha, 'hovercard')}
			>
				{sha.slice(0, 7)}
			</a>
		</code>
	);
}
