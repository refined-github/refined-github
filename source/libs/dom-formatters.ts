import zipTextNodes from 'zip-text-nodes';
import linkifyURLsCore from 'linkify-urls';
import linkifyIssuesCore from 'linkify-issues';
import getTextNodes from './get-text-nodes';
import {getOwnerAndRepo} from './utils';
import parseBackticksCore from './parse-backticks';

// Shared class necessary to avoid also shortening the links
export const linkifiedURLClass = 'rgh-linkified-code';

// If we are not in a repo, relative issue references won't make sense
// but `user`/`repo` need to be set to avoid breaking errors in `linkify-issues`
// https://github.com/sindresorhus/refined-github/issues/1305
const currentRepo = getOwnerAndRepo();
const options = {
	user: currentRepo.ownerName || '/',
	repo: currentRepo.repoName || '/',
	type: 'dom',
	baseUrl: '',
	attributes: {
		rel: 'noreferrer noopener',
		class: linkifiedURLClass // Necessary to avoid also shortening the links
	}
};

export function linkifyIssues(element: Element): void {
	const linkified = linkifyIssuesCore(element.textContent!, options);
	if (linkified.children.length === 0) { // Children are <a>
		return;
	}

	// Enable native issue title fetch
	for (const link of linkified.children as HTMLCollectionOf<HTMLAnchorElement>) {
		const issue = link.href.split('/').pop();
		link.setAttribute('class', 'issue-link js-issue-link tooltipped tooltipped-ne');
		link.setAttribute('data-error-text', 'Failed to load issue title');
		link.setAttribute('data-permission-text', 'Issue title is private');
		link.setAttribute('data-url', link.href);
		link.setAttribute('data-id', `rgh-issue-${issue}`);
	}

	zipTextNodes(element, linkified);
}

export function linkifyURLs(element: Element): void {
	if (element.textContent!.length < 15) { // Must be long enough for a URL
		return;
	}

	const linkified = linkifyURLsCore(element.textContent!, options);
	if (linkified.children.length === 0) { // Children are <a>
		return;
	}

	zipTextNodes(element, linkified);
}

export function parseBackticks(element: Element): void {
	for (const node of getTextNodes(element)) {
		const fragment = parseBackticksCore(node.textContent!);

		if (fragment.children.length > 0) {
			zipTextNodes(element, fragment);
		}
	}
}
