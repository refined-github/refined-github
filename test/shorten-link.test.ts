import {parseHTML} from 'linkedom';
import {expect, test} from 'vitest';

import {shortenLink} from '../source/github-helpers/dom-formatters.js';

function shortenLinksInFragment(html: string): string {
	const {document} = parseHTML(html);

	const links = document.querySelectorAll('a');
	for (const link of links) {
		shortenLink(link);
	}

	return document.documentElement.outerHTML;
}

test('shorten link in comment text', () => {
	// https://github.com/refined-github/refined-github/issues/4565#issue-943802539
	expect(shortenLinksInFragment(`
		<td class="d-block comment-body markdown-body js-comment-body">
			<p dir="auto">
				<a href="https://github.com/darkred/test/compare/main...t2?expand=1">https://github.com/darkred/test/compare/main...t2?expand=1</a>
			</p>
		</td>
	`)).toMatchSnapshot();
});

test('avoid shortening link in code block inside comment', () => {
	// https://github.com/denosaurs/mod.land/issues/55#issue-1160032701
	expect(shortenLinksInFragment(`
		<div class="highlight highlight-source-ts">
			<pre class="rgh-linkified-code">
				<span class="pl-s">
					<a href="https://raw.githubusercontent.com/denosaurs/mod.land/master/cnames.ts" rel="noreferrer noopener" class="rgh-linkified-code">
						https://raw.githubusercontent.com/denosaurs/mod.land/master/cnames.ts
					</a>
				</span>
			</pre>
		</div>
	`)).toMatchSnapshot();
});

test('avoid shortening link in embedded file preview inside comment', () => {
	// https://github.com/refined-github/refined-github/pull/4759#issue-988481591
	expect(shortenLinksInFragment(`
		<div class="comment-body markdown-body js-comment-body soft-wrap user-select-contain d-block">
			<div class="Box Box--condensed my-2">
				<div itemprop="text" class="Box-body p-0 blob-wrapper blob-wrapper-embedded data">
					<table class="highlight tab-size mb-0 js-file-line-container" data-tab-size="8" data-paste-markdown-skip="">
						<tbody>
							<tr class="border-0">
								<td id="LC45" class="blob-code border-0 px-3 py-0 color-bg-default blob-code-inner js-file-line rgh-linkified-code">
									<span class="pl-c">//
										<a href="https://github.com/sindresorhus/refined-github/issues/522#issuecomment-311271274" rel="noreferrer noopener" class="rgh-linkified-code">
											https://github.com/sindresorhus/refined-github/issues/522#issuecomment-311271274
										</a>
									</span>
								</td>
							</tr>
						</tbody>
					</table>
				</div>
			</div>
		</div>
	`)).toMatchSnapshot();
});

test('shorten link in review comment text', () => {
	// https://github.com/refined-github/refined-github/pull/4759#discussion_r738167140
	expect(shortenLinksInFragment(`
		<div class="comment-body markdown-body js-comment-body soft-wrap user-select-contain d-block">
			<p dir="auto">
				<a href="https://github.com/refined-github/refined-github">https://github.com/refined-github/refined-github</a>
			</p>
		</div>
	`)).toMatchSnapshot();
});

test('avoid shortening links in suggestion inside review comment', () => {
	// https://github.com/refined-github/refined-github/pull/4759#discussion_r738167140
	expect(shortenLinksInFragment(`
		<div class="comment-body markdown-body js-comment-body soft-wrap user-select-contain d-block">
			<div class="my-2 border rounded-2 js-suggested-changes-blob diff-view" id="">
				<div itemprop="text" class="blob-wrapper data file" style="margin: 0; border: none; overflow-y: visible; overflow-x: auto;">
					<table class="d-table tab-size mb-0 width-full" data-paste-markdown-skip="">
						<tbody>
							<tr class="border-0">
								<td class="border-0 px-2 py-1 blob-code-inner blob-code-addition js-blob-code-addition blob-code-marker-addition">
									https:<span class="pl-c">//github.com/refined-github/refined-github</span><span class="pl-kos"></span>
								</td>
							</tr>
						</tbody>
					</table>
				</div>
			</div>
		</div>
	`)).toMatchSnapshot();
});
