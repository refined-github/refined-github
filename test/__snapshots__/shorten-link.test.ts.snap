// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`avoid shortening link in code block inside comment 1`] = `
"<div class=\\"highlight highlight-source-ts\\">
			<pre class=\\"rgh-linkified-code\\">
				<span class=\\"pl-s\\">
					<a href=\\"https://raw.githubusercontent.com/denosaurs/mod.land/master/cnames.ts\\" rel=\\"noreferrer noopener\\" class=\\"rgh-linkified-code\\">
						https://raw.githubusercontent.com/denosaurs/mod.land/master/cnames.ts
					</a>
				</span>
			</pre>
		</div>"
`;

exports[`avoid shortening link in embedded file preview inside comment 1`] = `
"<div class=\\"comment-body markdown-body js-comment-body soft-wrap user-select-contain d-block\\">
			<div class=\\"Box Box--condensed my-2\\">
				<div itemprop=\\"text\\" class=\\"Box-body p-0 blob-wrapper blob-wrapper-embedded data\\">
					<table class=\\"highlight tab-size mb-0 js-file-line-container\\" data-tab-size=\\"8\\" data-paste-markdown-skip=\\"\\">
						<tbody>
							<tr class=\\"border-0\\">
								<td id=\\"LC45\\" class=\\"blob-code border-0 px-3 py-0 color-bg-default blob-code-inner js-file-line rgh-linkified-code\\">
									<span class=\\"pl-c\\">//
										<a href=\\"https://github.com/sindresorhus/refined-github/issues/522#issuecomment-311271274\\" rel=\\"noreferrer noopener\\" class=\\"rgh-linkified-code\\">
											https://github.com/sindresorhus/refined-github/issues/522#issuecomment-311271274
										</a>
									</span>
								</td>
							</tr>
						</tbody>
					</table>
				</div>
			</div>
		</div>"
`;

exports[`avoid shortening links in suggestion inside review comment 1`] = `
"<div class=\\"comment-body markdown-body js-comment-body soft-wrap user-select-contain d-block\\">
			<div class=\\"my-2 border rounded-2 js-suggested-changes-blob diff-view\\">
				<div itemprop=\\"text\\" class=\\"blob-wrapper data file\\" style=\\"margin: 0; border: none; overflow-y: visible; overflow-x: auto;\\">
					<table class=\\"d-table tab-size mb-0 width-full\\" data-paste-markdown-skip=\\"\\">
						<tbody>
							<tr class=\\"border-0\\">
								<td class=\\"border-0 px-2 py-1 blob-code-inner blob-code-addition js-blob-code-addition blob-code-marker-addition\\">
									https:<span class=\\"pl-c\\">//github.com/refined-github/refined-github</span><span class=\\"pl-kos\\"></span>
								</td>
							</tr>
						</tbody>
					</table>
				</div>
			</div>
		</div>"
`;

exports[`shorten link in comment text 1`] = `
"<td class=\\"d-block comment-body markdown-body js-comment-body\\">
			<p dir=\\"auto\\">
				<a href=\\"https://github.com/darkred/test/compare/main...t2?expand=1\\">darkred/test@<code>main...t2</code>?expand=1 (compare)</a>
			</p>
		</td>"
`;

exports[`shorten link in review comment text 1`] = `
"<div class=\\"comment-body markdown-body js-comment-body soft-wrap user-select-contain d-block\\">
			<p dir=\\"auto\\">
				<a href=\\"https://github.com/refined-github/refined-github\\">refined-github/refined-github</a>
			</p>
		</div>"
`;
